### ISSUE TITLE
API consideration with ruby-gtk: long-term guarantees for API stability irrespective of what upstream GTK devs change (randomly ...)

### ISSUE THREAD

**Issue by rubyFeedback** [ruby-gnome/ruby-gnome#1508]
Hey there kou and everyone else who may be using ruby-gtk or
perhaps has used ruby-gtk or plans on using it.

In the last days I played around a bit with ruby-gtk4. Things
work quite well, although a few things don't. Right now I
struggle with finding examples. Even python gtk4 examples are
rare to find; it seems there is not so much content for gtk4. When
I started with ruby-gtk3 more seriously a few years ago, the python
examples helped a LOT. I could almost translate them 1:1 onto
ruby-gtk3 and it worked for the most part. \o/

One thing I struggle with right now are ... events in gtk4.

In the past I could easily grab all events in ruby-gtk2 and
ruby-gtk3 on a given widget. So I had feedback on the widget, and then
after finding out the name of the event/signal on the commandline, I could
use it. Or I could look up in the old wiki, but that one is gone, and
would not be of any help for gtk4 anyway.

The documentation upstream for gtk4 is ... not that great.

I am a bit disappointed by the GTK dev team in that they
claim gtk4 is epic, but to me it almost looks as if they
did many meaningless changes that ... don't really improve
functionality. It's almost as if they change things for
the sake of changing things - like pretending it is now
when it is not. I am not claiming this is the case for ALL of
gtk, but to a subset of the changes in gtk4.

Meanwhile tons of old code is removed. I understand the
rationale and maintenance code, but I am beginning to
wonder WHY things are so quickly discarded - as if GTK
is solely used to cater to GNOME and ignore everyone
else.

Anyway.

In the past I could use:

    .signal_connect(:events)

Kind of, and then figure out the name inside a block.

I am not sure what can be used for ruby-gtk4 yet.

The migration guide they offer upstream is quite useless
to me without examples that work. If anyone wants to
add a few examples for ruby-gtk4 this would be great.

But this issue request is about something else.

Considering how upstream API constantly changes, I am
wondering whether it would not be better for ruby-gtk
itself to provide some kind of a "stable guarantee"
that people can use an "internal API", that is then
mapped onto whatever upstream may have changed.

I give you an example.

    add()    # works in ruby-gtk3
    append() # this is now how it has to be in ruby-gtk4

I don't see the point of the change. So I added

    add_or_append()

Or, simpler, to add a widget I may just use <<. That way
I can say "no matter what GTK devs do, I just use << and
it will work, for ruby-gtk2, ruby-gtk3, ruby-gtk4 and
so on". So I become less dependent on any particular API
changes upstream.

Rewriting ruby-gtk apps is a bit annoying. That was one
reason why I added a custom gem that does these translations
for me. And this is ok - I can use my custom gem. But!

I am also annoyed that I have to look up bad or missing
upstream documentation, fail at figuring out what changed
and then ... can not use it because I am too dumb to figure
it all out.

That's why I would like it if the ruby-gnome project itself
would add some kind of helper module that tries to guarantee
some stability in the API.

I am not saying all of the API should be guaranteed, mind
you!

I am saying that if an API change can be mapped with helper
code, then it would be better to use that, than rely to 
force people to adjust all the downstream code as-is.

I feel as if I am investing time largely now to satisfy
some random GTK upstream developers. And that's no fun.

So I'd like some kind of meta-API for ruby-gtk/ruby-gnome.

The glimmer project in some ways goes for that, but it
uses its own DSL, and that DSL is not simple to learn.
I don't necessarily want a DSL, I kind of want a stable
API to be able to use things - the add() versus append()
versus add_or_append() is an example of what I mean.

Or, with my latest frustration, events. I'd love to be
able to do something like this:

    window = Gtk::Window.new
    entry = Gtk::Entry.new
    window.add(entry)
    entry.signal_connect(:clicked) {|widget, event|
      pp event.class
    }

This kind of thing. I'd like this to work across all
ruby-gtk versions. Then I can use the appropriate name
for the event and go for that.

Or drag and drop. That has changed AGAIN. So now I have
to change all drag-and-drop code ...I can write wrapper
code, but I often don't manage to figure out what is
going on, and the documentation is not great.

Anyway.

This is a bit more of a frustration and what not. It's
ok to close this issue. I kind of hope that I can speak
a bit for other ruby-devs who find GUI work a bit frustrating,
and would like for things to become a bit simpler and more
transparent, rather than depend on whatever random upstream
change is made by gtk. They are already talking about GTK5
and I bet many more things will again change ... it's so
frustrating to change and adjust. Right now ruby-gtk3 is
far easier to use than ruby-gtk4. Perhaps in the future this
may change but right now I'll stick with ruby-gtk3.

TL;DR: if anyone can figure out event handling in ruby-gtk4
it would be super-appreciated. For now I am giving up on
it and go back to ruby-gtk3 - there things are kind of easier
for the time being.

**Comment by kojix2** [comment]
Why not create pull requests to take that frustration out of your mind?
If you don't submit code, nobody will write it, right?

**Comment by kou** [comment]
We can add some gtk4 gem compatible API to gtk3 gem (for example, we can add `append` to gtk3 gem) but this issue is too complicated to figure what should we do for smooth gtk3 gem -> gtk4 gem migration. So I close this for now.

We can create some small issues for gtk3 gem -> gtk4 gem migration instead.
